woocommerce/woocommerce
# [WIP] CYS - Core: add integration with the font library

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes #41924.

This PR implements the support for the Font Library to allow the user to pick different fonts on the Assembler.

## Technical Details

### Font Library API

The Font Library is based on two post types after the [API refactor](https://github.com/WordPress/gutenberg/pull/57688): 
`wp_font_family` -> contains the data related to the Font Family (e.g.: slug)
`wp_font_face` -> includes the data of specific font faces of a particular font family.

There is a strict relationship between these two post types. The `wp_font_family` is the parent post of the `wp_font_face`. For this reason,  it is necessary to create a `wp_font_family`post before installing the various font faces. 

### CYS integration

The fonts installed are defined with the `FONT_FAMILY_TO_INSTALL` constant. 

This constant is created manually by reading the value from the `FONT_PAIRINGS_WHEN_AI_IS_OFFLINE` constant ([source code](https://github.com/woocommerce/woocommerce/blob/6cfe61d2f877329ab02d2a7001125bf4e7fdba91/plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/global-styles/font-pairing-variations/constants.ts/#L13-L74)): we want to install only the necessary fonts to reduce the setup time and not install too many fonts. We could create a logic to generate this variable, but at this time, it isn't worth it, in my opinion.

During the initial setup, the client gets the already installed Font Families (and related Font Face). In this way, we can install only the fonts that aren't already installed. ([source code](https://github.com/woocommerce/woocommerce/blob/6cfe61d2f877329ab02d2a7001125bf4e7fdba91/plugins/woocommerce-admin/client/customize-store/design-without-ai/services.ts/#L91-L97)) 


After the setup is finished, when the Assembler is loaded, all the fonts are preloaded. ([source code](https://github.com/woocommerce/woocommerce/blob/6cfe61d2f877329ab02d2a7001125bf4e7fdba91/plugins/woocommerce-admin/client/customize-store/assembler-hub/preload-fonts.tsx)) 







### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new WooCommerce installation with this version.
2. Ensure the WooCommerce Beta Tester plugin is installed and activated (available on this monorepo).
3. Head over to `/wp-admin/tools.php?page=woocommerce-admin-test-helper` and enable `customize-store` feature flag.
4. Install this build of Gutenberg that contains the new Font Library API: 
[gutenberg.zip](https://github.com/woocommerce/woocommerce/files/14085546/gutenberg.zip)


5. Visit the `wp-admin/admin.php?page=wc-admin&path=/customize-store`.
6. Open the network tools.
7. Follow the process.
8. Ensure that there are several POST requests to the `/wp/v2/font-families/` endpoint.
9. Click "Change fonts" on the left sidebar.
10. Select different fonts and ensure that the selected font is applied to the first title
11. Visit the `wp-admin/admin.php?page=wc-admin&path=/customize-store`.
12. Open the network tools.
13. Ensure that no POST requests to the `/wp/v2/font-families/` endpoint are done.

![image](https://github.com/woocommerce/woocommerce/assets/4463174/62cb8349-5939-433f-bb37-c1fee1d2e80e)

### Warnings ⚠️ 

- Currently, in some cases, the font assigned to the buttons isn't the correct one #44010.
- Some E2E tests are failing. The reason is that the Font Library API isn't available on the E2E tests env. Before merging this PR, it is necessary to implement a way to detect if the Font Library isn't available to avoid fetching requests to non-existent endpoints. This work will be done in a dedicated PR.





### Changelog entry




<details>

#### Significance




#### Type




#### Message 

CYS - Core: add integration with the font library.

#### Comment 

</details>
